maintenance/projects/js-toolkit/packages/liferay-npm-build-support/src/loader/normalize-angular-output.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: LGPL-3.0-or-later
+ */
+
+import {
+	BundlerLoaderContext,
+	BundlerLoaderReturn,
+} from 'liferay-npm-build-tools-common/lib/api/loaders';
+
+/**
+ * This loader:
+ *
+ *   1. Discards any webpack artifact not containing ES5 code
+ *   2. Removes `-es5` from file and module names when present
+ *
+ * This way, the bundler can treat any known Angular version the same, no matter
+ * how webpack produced its output.
+ *
+ * This is because Angular version <=10 produce artifacts for ES5 and ES2015
+ * targets, while Angular versions >=10 only produce one type of artifact, not
+ * differentiated by file name.
+ */
+export default function (context: BundlerLoaderContext): BundlerLoaderReturn {
+	const {filePath, log} = context;
+
+	if (filePath.includes('-es') && !filePath.includes('-es5.')) {
+		context.content = undefined;
+		context.extraArtifacts = {};
+
+		log.info(
+			'normalize-angular-output',
+			`Discarding output because source is not ES5 code`
+		);
+
+		return undefined;
+	}
+
+	if (filePath.includes('-es5.') && context.content) {
+		const newFilePath = filePath.replace('-es5.', '.');
+
+		context.extraArtifacts[newFilePath] = context.content.replace(
+			'-es5',
+			''
+		);
+		context.content = undefined;
+
+		log.info(
+			'normalize-angular-output',
+			`Diverted main output to ${newFilePath}`
+		);
+	}
+
+	Object.keys(context.extraArtifacts).forEach((filePath) => {
+		if (filePath.includes('-es5.') && context.extraArtifacts[filePath]) {
+			const newFilePath = filePath.replace('-es5.', '.');
+
+			context.extraArtifacts[newFilePath] = context.extraArtifacts[
+				filePath
+			].replace('-es5', '');
+			delete context.extraArtifacts[filePath];
+
+			log.info(
+				'normalize-angular-output',
+				`Diverted extra artifact ${filePath} to ${newFilePath}`
+			);
+		}
+	});
+
+	return undefined;
```

Comment:
![too-much](https://user-images.githubusercontent.com/7074/110772007-731f5480-825b-11eb-9700-4180286eedc5.gif)

Reply:
It is required by `tsc` because the  type is defined as `undefined|something`.

Reply:
Really? You don't even [have strict mode on](https://github.com/liferay/liferay-frontend-projects/blob/abe0815c66bc109763902f2ec9b029770ba6c721/maintenance/projects/js-toolkit/tsconfig.json), and AFAIK, not even [strict mode](https://www.typescriptlang.org/tsconfig#strict) will complain about this by default (you have to explicitly turn on `noImplicitReturns`, which is [off by default](https://www.typescriptlang.org/tsconfig#noImplicitReturns)).

ie. paste this into the [TS playground](https://www.typescriptlang.org/play?noImplicitReturns=false&ssl=1&ssc=1&pln=3&pc=2#code/GYVwdgxgLglg9mABADwBQEoBciBGc4A2ApgIZIA+i4AJkcDGEdYgN4BQiniMwiqAsiSgALAHQAnMtTgBbDIgB8iUQCZ0icUSghxSKOJBEA3GwC+QA), ensuring that `noImplicitReturns` is off:

```ts
function x(): boolean | undefined {
    if (Math.random() > .2) return true;
}
```

Reply:
I think the reason the default is off is because the "philosophy" of TS (per [their design goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals)) is to allow people to write "natural, idiomatic" JS, and not writing `return undefined;` is an example of that.